### ISSUE TITLE
Seems unable to convert <pre class="EnlighterJSRAW"> </pre>to Markdown code block

### ISSUE THREAD

**Issue by A-Tan5** [deathau/markdownload#69]
Hi, 
thanks for this amazing tool.

I was trying to convert [this page](https://www.appcoda.com.tw/functional-reactive-programming/) to Markdown,
and I was using markdownload on Brave browser _(version 1.21.77, Chromium: 89.0.4389.90 )_ on mac _(Macbook pro 2020)_.

Everything worked fine, but the content inside `<pre class="EnlighterJSRAW">` and `</pre>` turns out to be plain text, which should be code block base on the website appearance.

I didn't see any setting option related to this situation, so I thought maybe creating an issue would be helpful. Thanks!

**Comment by Runskey** [comment]
+1! I also found that tag `pre` is not converted to code block. Wish you can consider it.

**Comment by christophberger** [comment]
I, too, would love to see the `pre` tag being processed. An example for the pre tag can be found here: 

https://go.googlesource.com/proposal/+/refs/heads/master/design/43651-type-parameters.md

The code on this page is enclosed in `<pre class="code">` tags. Example:

```html
<pre class="code">// Print prints the elements of a slice.
// It should be possible to call this with any slice value.
func Print(s []T) { // Just an example, not the suggested syntax.
	for _, v := range s {
		fmt.Println(v)
	}
}
</pre>
```
The downloaded MD page shows this text as plain text, without any renderable line endings or indenting.

Additionally, markdownload adds escape sequences, like `\[` or `\_`, which makes it more tedious to turn the text back into code.

```
// Print prints the elements of a slice.
// It should be possible to call this with any slice value.
func Print(s \[\]T) { // Just an example, not the suggested syntax.
for \_, v := range s {
fmt.Println(v)
}
}
```
When rendered (e.g. in Obsidian), the text even has no line endings anymore.

> // Print prints the elements of a slice. // It should be possible to call this with any slice value. func Print(s []T) { // Just an example, not the suggested syntax. for _, v := range s { fmt.Println(v) } }

It would be super helpful on code-intense pages if markdownload turns `pre` tags into code blocks.
Thanks!

**Comment by christophberger** [comment]
Looks like this is a problem of the Turndown library that requires `<pre>` to have a child element `<code>`. Alas, not all HTML pages stick to that rule. 

See 

https://github.com/deathau/markdownload/blob/63e27859c5eef8614b87d0e4475396bfdf22c9c5/src/background/turndown.js#L149-L157

and 
https://github.com/deathau/markdownload/blob/63e27859c5eef8614b87d0e4475396bfdf22c9c5/src/background/turndown.js#L168-L176

I am neither a JavaScript expert nor a Turndown expert, but might a [custom rule](https://github.com/mixmark-io/turndown#extending-with-rules) help to relax the requirement to only `<pre>`?

**Comment by deathau** [comment]
I've added some support to parse `pre` elements as code blocks now, and if it's got the same classes as GitHub it will put in the language for syntax highlighting, too.
Because of all the different syntax highlighting libraries out there, I'm only supporting GitHub for now, but I might add more later.

Some caveats:
@A-Tan5 — for some reason, in the link you provided, the extension will output some unfenced code, followed by the correctly fenced code. This is to do with the way the syntax highlighter is outputting everything on that page not being able to be parsed easily by Readability, I think. But there's also a hidden `pre` that works great, so there is fenced code _as well_.

@christophberger — the page you linked to will parse the `pre`s as code blocks but for some reason loses all the indentation. Not sure why, and it doesn't happen on other pages I tested.
